io/io/src/TBufferMerger.cxx

```diff
@@ -104,6 +109,8 @@ void TBufferMerger::WriteOutputFile()
       merger.OutputFile(std::unique_ptr<TFile>(fFile));
    }
 
+   auto mergefn = [&merger] { merger.PartialMerge(); merger.Reset(); };
```

Comment:
The callback mechanism is meant to be used by the user, not the merger class itself. If you use this mechanism in the merger, the user can then overwrite `fCallback` by registering their own callback function, and all data will be thrown away instead of being merged to the output file, as the user cannot access the merger in their callback function.

Reply:
This uses nested callbacks--the user callback gets the `mergefn` as a callback, which it needs to call to execute the merge.  The `mergefn` lambda "captures" the merger, so the user callback doesn't need access to it, it just needs to call the lambda that captured the merger.

Reply:
> the user callback gets the `mergefn` as a callback, *which it needs to call to execute the merge*

This is the part that worries me. I think that we should not create a mechanism that let's the user disable merging altogether, especially because this is the default behaviour if, for example, a callback that just prints the size of the queue is registered.